### ISSUE TITLE
[Bug]: weaviate client has no collection attributes

### ISSUE THREAD

**Issue by arsyad2281** [run-llama/llama_index#13614]
### Bug Description

I am initializing `WeaviateVectorStore` based on the example given in this [link](https://docs.llamaindex.ai/en/stable/api_reference/storage/vector_store/weaviate/?h=weaviatevect#llama_index.vector_stores.weaviate.WeaviateVectorStore)

However, I am getting this error:
`AttributeError: 'Client' object has no attribute 'collections'`

Python Version - 3.11.9

Version of llama index and relevant packages
llama-index==0.10.37
llama-index-vector-stores-weaviate==1.0.0

### Version

0.10.37

### Steps to Reproduce

Please replace `<your-username>`, `<your-password>` and `<your-index-name>` accordingly.
WeaviateDB is locally hosted through docker with this image: `semitechnologies/weaviate:1.23.9`

```
from llama_index.vector_stores.weaviate import WeaviateVectorStore
import weaviate

weaviate_url = "http://localhost:8080"

resource_owner_config = weaviate.AuthClientPassword(
    username=<your-username>,
    password=<your-password>,
)

client = weaviate.Client(
    weaviate_url,
    auth_client_secret=resource_owner_config,
)

vector_store = WeaviateVectorStore(
    weaviate_client=client, index_name=<your-index-name>
)
```

### Relevant Logs/Tracbacks

```shell
AttributeError                            Traceback (most recent call last)
Cell In[22], line 1
----> 1 vector_store = WeaviateVectorStore(
      2     weaviate_client=client, index_name="Ccassistant"
      3 )

File ~/homebrew/Caskroom/miniconda/base/envs/test_env/lib/python3.11/site-packages/llama_index/vector_stores/weaviate/base.py:172, in WeaviateVectorStore.__init__(self, weaviate_client, class_prefix, index_name, text_key, auth_config, client_kwargs, url, **kwargs)
    167     raise ValueError(
    168         "Index name must start with a capital letter, e.g. 'LlamaIndex'"
    169     )
    171 # create default schema if does not exist
--> 172 if not class_schema_exists(self._client, index_name):
    173     create_default_schema(self._client, index_name)
    175 super().__init__(
    176     url=url,
    177     index_name=index_name,
   (...)
    180     client_kwargs=client_kwargs or {},
    181 )

File ~/homebrew/Caskroom/miniconda/base/envs/test_env/lib/python3.11/site-packages/llama_index/vector_stores/weaviate/utils.py:76, in class_schema_exists(client, class_name)
     74 """Check if class schema exists."""
     75 validate_client(client)
---> 76 return client.collections.exists(class_name)

AttributeError: 'Client' object has no attribute 'collections'
```

**Comment by arsyad2281** [comment]
Do note that collection is also called in this function - [link](https://github.com/run-llama/llama_index/blob/8f1e97874f016573e5b4a36c215d9547c8f3526d/llama-index-integrations/vector_stores/llama-index-vector-stores-weaviate/llama_index/vector_stores/weaviate/base.py#L291), where `collection` is expected to be an object in L328 :

```
def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResult:
        """Query index for top k most similar nodes."""
        all_properties = get_all_properties(self._client, self.index_name)
        collection = self._client.collections.get(self.index_name)
        filters = None
        ...
        ...
        # execute query
        try:
            query_result = collection.query.hybrid(
                query=query.query_str,
                vector=vector,
                alpha=alpha,
                limit=limit,
                filters=filters,
                return_metadata=return_metatada,
                return_properties=all_properties,
                include_vector=True,
            )
        except weaviate.exceptions.WeaviateQueryError as e:
            raise ValueError(f"Invalid query, got errors: {e.message}")
        ...
        ...
```

**Comment by chrisk314** [comment]
@arsyad2281 from your issue description, it looks like you made the same mistake I did. Change your code as per my comment above and you should be good.

**Comment by chrisk314** [comment]
On closer inspection I realised that the `WeaviateVectorStore.from_params` method was creating a Weaviate V3 client, where the rest of the code is expecting a Weaviate V4 client. I made a [PR](https://github.com/run-llama/llama_index/pull/13719) which changes all the code to require and use Weaviate V4 clients. This choice makes the updated code consistent and error free with respect to the rest of the existing code; and ensures that the code is future proofed by migrating to Weaviate V4 now.

**Comment by krisz094** [comment]
I just ran into a similar issue, where I try to ingest modified files into my Weaviate database, so basically doing an upsert.

I think it's caused by the `base.py` code using mixed V3 and V4 code too.

The error I get:

```bash
Traceback (most recent call last):
  File "/foo/bar.py", line 56, in <module>
    nodes = pipeline.run(documents=documents)
  File "/home/krisz/.pyenv/versions/3.10.13/lib/python3.10/site-packages/llama_index/core/ingestion/pipeline.py", line 682, in run
    nodes_to_run = self._handle_upserts(
  File "/home/krisz/.pyenv/versions/3.10.13/lib/python3.10/site-packages/llama_index/core/ingestion/pipeline.py", line 612, in _handle_upserts
    self.vector_store.delete(ref_doc_id)
  File "/home/krisz/.pyenv/versions/3.10.13/lib/python3.10/site-packages/llama_index/vector_stores/weaviate/base.py", line 261, in delete
    self._client.query.get(self.index_name)
AttributeError: 'WeaviateClient' object has no attribute 'query'
```

The relevant part in `base.py`:

```python
query = (
            self._client.query.get(self.index_name)
            .with_additional(["id"])
            .with_where(where_filter)
            .with_limit(10000)  # 10,000 is the max weaviate can fetch
        )
```

The `_client.query` doesn't exist in V4 anymore, I think.

**Comment by chrisk314** [comment]
@krisz094 thanks for pointing that out. I've pushed a commit dc8f15e to #13719 which fixes the `WeaviateVectorStore.delete` method. Just tested manually and it's working now. I didn't test it out with additional `delete_kwargs["filters"]` yet.

Weaviate and LlamaIndex code bases are both very new to me; I've based my fix for the `delete` method off code that's in the `query` method and what I can piece together from the Weaviate code and docs. I'm pushed for time right now and it's a long weekend in the UK. I'll be able to look more in depth at the rest of the `WeaviateVectorStore` code on Tuesday and test further.

**Comment by chrisk314** [comment]
Just following up on this... I've now manually tested all public methods of the `WeaviateVectorStore` with the updates in the PR and all looks good to me.

Any timeline for a review from any of the core contributors?

**Comment by brenkehoe** [comment]
#13365 should fix delete method @chrisk314

**Comment by chrisk314** [comment]
@brenkehoe ok thanks for pointing that out. I've now merged the latest changes from main into #13719  taking the `delete` implementation from main.
